Implement Base Level Design

knittingarch opened the pull request:
This PR addresses the challenges of using Gatsby for the first time with basic React skills.  I was able to accomplish many of the objectives of the blog setup, but there are still some pieces that aren't quite working as planned.

All in all, though, I think the use of components has made updating the design easier in the future, and I also expect it to help as the site becomes more complex with search or other functional extensions.

knittingarch committed:
More Design Changes

I'm still trying to figure out the fonts which aren't being imported properly.  Additionally, I'm having trouble with some of the positioning and layout on the index page.

Next moves are to add an About page so that I can fix that link on the index.  I also need to add a preview field in contentful and use that in the Featured Post.  I also want to try fixing the images as they aren't displaying in an aesthetically pleasing way.  Possibly needs a component?  Styled components are a bit tricky for this stuff.  Finally, I want to revisit the font issues as that will help the look and feel of the page.

Remove Strange Font Import Setup

I tried to import fonts on a few ways that didn't work.  After consulting the official Gatsby docs, I discovered that they recommend a global.css file and importing fonts from GoogleFonts directly.  Once I got that working, removing the other methods was quick and easy.

Refactor Blog Post Into Various Implementations

While it appeared in the beginning that there was only one use case for the BlogPost component, it became clear as I continued to build and brainstorm that there were actually three use cases.  So, I have created a component for the following cases: featured on the index page, full post on its own page, and simplified for the All Posts page.

Now I am realizing that there is a lot of code that has been duplicated.  So it's time to create some additional components to streamline updates.  I'll be starting with the following: Date, Tags, Title, Image, Paragraph, TruncatedParagraph, and Links.  May need more later... LOL

Reduce Duplication With Components

This commit implements the changes prescribed in the last commit.  There is an issue with tiling the pictures for the individual blog post, but hopefully I can figure that out.

The only other page that needs imaginging is the About This Blog page.  Not quite sure what I want that page to look like.  I also need to add a horizontal rule between the posts on the All Posts page.

knittingarch committed:
More Design Changes

I'm still trying to figure out the fonts which aren't being imported properly.  Additionally, I'm having trouble with some of the positioning and layout on the index page.

Next moves are to add an About page so that I can fix that link on the index.  I also need to add a preview field in contentful and use that in the Featured Post.  I also want to try fixing the images as they aren't displaying in an aesthetically pleasing way.  Possibly needs a component?  Styled components are a bit tricky for this stuff.  Finally, I want to revisit the font issues as that will help the look and feel of the page.

Remove Strange Font Import Setup

I tried to import fonts on a few ways that didn't work.  After consulting the official Gatsby docs, I discovered that they recommend a global.css file and importing fonts from GoogleFonts directly.  Once I got that working, removing the other methods was quick and easy.

Refactor Blog Post Into Various Implementations

While it appeared in the beginning that there was only one use case for the BlogPost component, it became clear as I continued to build and brainstorm that there were actually three use cases.  So, I have created a component for the following cases: featured on the index page, full post on its own page, and simplified for the All Posts page.

Now I am realizing that there is a lot of code that has been duplicated.  So it's time to create some additional components to streamline updates.  I'll be starting with the following: Date, Tags, Title, Image, Paragraph, TruncatedParagraph, and Links.  May need more later... LOL

Reduce Duplication With Components

This commit implements the changes prescribed in the last commit.  There is an issue with tiling the pictures for the individual blog post, but hopefully I can figure that out.

The only other page that needs imaginging is the About This Blog page.  Not quite sure what I want that page to look like.  I also need to add a horizontal rule between the posts on the All Posts page.

dimanyc commented on src/components/BlogPost.js:
```diff
@@ -1,44 +1,41 @@
+    <Title title={post.title} />
+    <ImageWrapper>
+      {post.media.map((image) => (
+          <Image
```
`<Image />` might need a a key prop https://reactjs.org/docs/lists-and-keys.html

knittingarch replied:
Ah, yes.  Good catch!  I think I added a key for the original BlogPost component, but forgot about this one (and others) when I made it during a refactor :)

dimanyc commented on src/components/FeaturedBlogPost.js:
```diff
@@ -0,0 +1,35 @@
+  <BlogPostWrapper>
+    <Date date={post.publishedDate}/>
+    {post.tags.map((tag) => (
+      <Tag tag={tag} />
```
same: `<Tag />` might need a a key prop https://reactjs.org/docs/lists-and-keys.html

dimanyc commented on src/components/FeaturedImage.js:
```diff
@@ -0,0 +1,20 @@
+
+const Image = ({ src, alt }) => (
+  <BlogPostImageWrapper>
+      <img src={src} alt={alt} style={ImageWidth} />
```
I think the indentation is off here

dimanyc commented on src/pages/about.js:
```diff
@@ -0,0 +1,41 @@
+  marginTop: '20px',
+}
+
+class AboutPage extends React.Component {
```
What do you think about making this a functional component? Looks like `AboutPage` is not using an internal state anyway.

dimanyc commented on src/pages/blog-posts.js:
```diff
@@ -13,12 +26,11 @@ class BlogPosts extends React.Component {
-            <BlogPost key={post.node.slug} post={post.node} />
-            <Link to={`/${post.node.slug}`}>{post.node.title}</Link>
-            <br />
+            <SimplifiedBlogPost key={post.node.slug} post={post.node} />
```
same here. Could this be a functional component?

knittingarch replied:
This refactor didn't work.  Having posts inside the functional component threw an error, but so did moving it outside the component.  Not sure what the proper syntax would be to get this working as an FC.  Could we look at it Friday if you have time :)

dimanyc replied:
of course!

dimanyc commented on src/components/BlogPost.js:
```diff
@@ -1,44 +1,41 @@
+import Tag from './Tag'
+import Title from './Title'
+
+// Possible Colors
```
do you still need this comment here? I noticed that color declarations were removed here.

dimanyc reviewed (commented):
Great job!!! :rocket: Left a couple of comments with my thoughts. Quick note, I noticed a couple of styling inconsistencies but I also see that you have `prettier` added. I think you can just set your code editor to run prettier on save or something of that sort. This way you won't have to deal with styling issues ever again! 

Another great lib I can recommend is eslint: https://eslint.org/. Prettier and ESLint are usually used together to find and fix linting issues automatically. 

Happy to pair on this if you want.

knittingarch commented:
@dimanyc Thanks for the thorough review!  I have used eslint a lot, but didn't realize it wasn't installed running in my editor until now... LOL I think it's because I started sketching this out on my work computer which does have everything set up...  I should be able to get that sorted today, but if I have trouble, I'll ping you for a pair :)  I also thought I had prettier and Hound set up.  I guess I got a little ahead of myself :)

knittingarch committed:
Clean Up Code

This commit addresses PR comments.  The work done includes adding keys where needed, adding Prettier and eslint, adding proptypes, and alphabetizing imports.  The code looks much cleaner now and will be easier to test and augment going forward.

knittingarch committed:
Clean Up Code

This commit addresses PR comments.  The work done includes adding keys where needed, adding Prettier and eslint, adding proptypes, and alphabetizing imports.  The code looks much cleaner now and will be easier to test and augment going forward.

knittingarch commented:
@dimanyc I've made all the changes you requested (except that one refactor).  I also added proptypes and did some other cleanup like alphabetizing the imports and making sure the component names were consistent. Would you mind officially reviewing this PR?

dimanyc commented on src/templates/blog-post.js:
```diff
@@ -1,18 +1,25 @@
-        <BlogPost key={post.slug} post={post} />
-        <Link to="/">Back to Home</Link>
+      <div>
+        <BlogPost key={post.id} post={post} />
```
why use a key here?

knittingarch replied:
Missed during refactoring.

dimanyc commented on src/pages/index.js:
```diff
@@ -1,60 +1,64 @@
-            key={post.slug}
-            post={post}
-          />
+          <FeaturedBlogPost key={post.slug} post={post} />
```
no need for a `key` prop here. Key is used by React to determine changes in state when you are iterating over an array:

```jsx
['foo', 'bar', baz'].map((str) => <div key={str}>{str}</div> 
```
`key` prop here tells react when it needs to re-render. 

In this case `<FeaturedBlogPost />` component is static and it "exists" in React's virtual dom by default.

Let me know if this makes sense. Happy to explain this in person

knittingarch replied:
Missed during refactoring.

dimanyc commented on src/pages/blog-posts.js:
```diff
@@ -1,41 +1,51 @@
+  return (
+    <div>
+      {posts.map(post => (
+        <>
```
so this will raise a error because you are only adding a key to on elem -- `SimplifiedBlogPost`, whereas others are left without a key. I'd recommend wrapping this entire group into a div instead of a fragment `</>` and adding a key to it. What do you think?

knittingarch replied:
Checked the console, and you are correct errors abound!  Updating this :) Thanks for the heads up!

knittingarch commented on src/components/BlogPost.js:
```diff
@@ -23,20 +20,21 @@ const ImageWrapper = styled.div`
-    {post.tags.map((tag) => (
-      <Tag tag={tag} />
+    {post.tags.map((tag, index) => (
+      <Tag key={index} tag={tag} />
```
@dimanyc So I understand why I need keys and that they should be unique, but my tags don't have an ID of their own.  A few people suggested that I use an index, but then I found [this article](https://medium.com/@robinpokorny/index-as-a-key-is-an-anti-pattern-e0349aece318), so I'm thinking this isn't the best option.  I do plan to sort posts based on tags in the future, but I'm not sure if that is enough of a reason to avoid using an index...  I looked at using a [shortid](https://www.npmjs.com/package/shortid), but since my tags are created in Contentful, I don't see a way to generate them upon creation, and generating them in the component would mean they would change each page render, right? Do you have any suggestions on this front?

knittingarch committed:
Merge branch 'add-design-elements' of https://github.com/knittingarch/AnxiousInBrooklyn into add-design-elements

dimanyc commented on src/components/BlogPost.js:
```diff
@@ -1,45 +1,40 @@
+    ))}
+    <Title title={post.title} />
+    <ImageWrapper>
+      {post.media.map((image, index) => (
```
small nit: instead of using an index for key, you can just use `image.file.url` since that value is always unique. I believe eslint raises a warning about this.

knittingarch replied:
Yes, I posted a long comment here about indexes.  Missed that I had used one here, though.  Thanks for the catch!

dimanyc reviewed (approved):
lgtm. One small nit about using index in keys. Feel free to ignore
